keps/sig-network/2365-ingressclass-namespaced-params/README.md

```diff
@@ -0,0 +1,255 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This KEP proposes adding an optional Namespace field to the IngressClass
+ParametersRef field.
+
+## Motivation
+
+After the initial release of IngressClass, a number of use cases called for the
+ability to reference namespace-scoped Parameters. For example, one could use a
+GatewayClass parameters CR to describe how and where a controller should be
+provisioned. This same thought process was also happening in the Service APIs
+subproject. It was ultimately deemed worthwhile for GatewayClass if we could
+also gain approval for a parallel API change to IngressClass.
+
+### Goals
+
+- Allow referencing namespace-scoped Parameters resources.
+
+### Non-Goals
+
+- Requiring all Parameters resources to be namespace-scoped.
+
+## Proposal
+
+Add an optional Namespace field to the IngressClass ParametersRef field.
+
+### Risks and Mitigations
+
+The option to reference namespace-scoped Parameters resources could lead to
+confusion. It is relatively rare for resource references to be able to target
+both cluster-scoped and namespace-scoped resources. We believe that the
+advantages of this KEP outweigh this potential confusion.
+
+## Design Details
+
+This will result in adding a new `IngressClassParametersReference` type that
+closely mirrors the existing `TypedLocalObjectReference` type that is currently
+in use.
+
+```golang
+// IngressClassParametersReference identifies an API object. This can be used
+// to specify a cluster-scoped or namespace-scoped resource.
+type IngressClassParametersReference struct {
+  // APIGroup is the group for the resource being referenced. If APIGroup is not
+  // specified, the specified Kind must be in the core API group. For any other
+  // third-party types, APIGroup is required.
+	// +optional
+	APIGroup *string
+	// Kind is the type of resource being referenced.
+	Kind string
+	// Name is the name of resource being referenced.
+  Name string
+  // Namespace is the namespace of the resource being referenced. When empty,
+  // this reference will refer to a cluster-scoped resource.
+  // +optional
+  Namespace string
```

Comment:
Does this need to be a pointer?

Reply:
My understanding was that we only need a pointer when there is a distinction between a nil and empty value. In this case "" and nil would be interpreted in the same way. Wherever possible I try to avoid pointers because they can be a pain to work with. We can use `omitempty` to ensure that "" would be omitted from serialization here. This matches the approach we've taken with service-apis which I believe was inspired by upstream.

Reply:
No.  Convention is that all optional fields should be pointers.  Yes, there's an optimization we *could* (and sometimes have) made when the Go zero-value is not a valid value or is the correct default, but until we have machinery that automates that (i.e. and IDL) I'd rather stick to convention, unless we have a good reason (e.g. scale) not to.

Reply:
The confusion of `+optional` and pointers strikes again. 😁

Reply:
@apelisse has spent time on this problem and @DirectXMan12 is thinking around IDL. No action needed, but something to think about - if we have an IDL should we make that optimization in all possible cases?

Reply:
How broadly should we be applying this pattern? Should we adjust our types in service-apis as well to use pointers for all optional fields, even if nil and zero values are treated identically?

Reply:
Please don't have an unset field mean cluster scoped. I recommend an explicit "scope: cluster/namespace" in addition to having the namespace field.

Unset fields are best used to indicate "I don't have an opinion". Our API sadly has lots of anti-examples.

Reply:
Good call, I've expanded this to add a scope field as well. Is it still appropriate to consider the namespace field optional if it will be required when scope==cluster?